### DISCUSSION TITLE
Native Code Integration

### DISCUSSION THREAD

**Discussion by satzguyz** [nasa/openmct#5080] [Community Q&A]
Hello

I have a native library (written in C++) for interfacing the Telemetry data and Groundstation controller. Is there a way to integrate it directly or I need convert everything into *.js plugins.

**Comment by ja51d** [comment]
You don't need to convert your C++ code into a Javascript plugin. Open MCT is designed to be 'data agnostic,' meaning it doesn't care what language your back-end is written in as long as it can communicate over standard web protocols.

The standard architecture for this is:

C++ Middleware: Keep your C++ library as a standalone service. Wrap it in a simple WebSocket or HTTP server (using something like Crow, uWebSockets, or Boost.Beast).

Telemetry Adapter: Create a small Open MCT plugin (in JS) that acts as a client. It will subscribe to your C++ server's stream.

Integration: Use the openmct.telemetry.addProvider API to map the data coming from your C++ library into Open MCT.

This way, your C++ library handles the heavy lifting of the telemetry data, and Open MCT just handles the visualization.
